source/_posts/2016-07-09-exporting-processing-data.markdown

```diff
@@ -0,0 +1,116 @@
+
+If you are just curious what's stored in your database then you can use the `sqlite3` command-line tool or a graphical one like [DB Browser for SQLite](http://sqlitebrowser.org/).
+
+The table that is holding the states is called `states`. The `events` tables is responsible for storing the events which occurred. So, we will first check how many entries there are in the `states` table. `sqlite3` needs to know where the databases is located. To work with your database make sure that Home Assistant is not running or create a copy of the existing database. It's recommended to work with a copy.
+
+```bash
+$ sqlite3 /home/ha/.homeassistant/home-assistant_v2.db 
+SQLite version 3.11.0 2016-02-15 17:29:24
+sqlite> SELECT count(*) FROM states;
+24659
+```
+
+Let's have a look at a sample [SQL](https://en.wikipedia.org/wiki/SQL) query. This query will show all states in a period for the sensor `sensor.aare`.
+
+```sql
+SELECT state, last_changed FROM states
+  WHERE
+    entity_id = 'sensor.aare'
+  AND
+     last_changed BETWEEN
+    '2016-07-05 00:00:00.000000' AND '2016-07-07 00:00:00.000000';
+```
+
+The SQL statement can be formed that it fits exactly what you need. This means that you can process the data in any way you want for further use. Often it makes sense to eliminate certain entries like `Unknown` or peaks.
+
+If the above query is executed in DB Browser for SQLite you would be able to save the sensor's graph as png.
+
+<p class='img'>
+  <img src='/images/blog/2016-07-reporting/db-browser.png' />
+  Visualization with DB Browser for SQLite
+</p>
+
+You may ask: Why not do this with LibreOffice Calc or another spreadsheet application? As most spreadsheet applications are not able to work directly with SQLite database we are going to export the data from the database to [CSV](https://en.wikipedia.org/wiki/Comma-separated_values).
+
+```bash
+$ sqlite3 -header -csv /home/ha/.homeassistant/home-assistant_v2.db "SELECT last_changed, state FROM states WHERE entity_id = 'sensor.aare' AND last_changed BETWEEN '2016-07-05 00:00:00.000000' AND '2016-07-07 00:00:00.000000';" > sensor.csv
+```
+
+The ordering for the `SELECT` was changed to get the time stamps first and then the state. Now we can import the CSV file into the application of your choice, here it's LibreOffice Calc.
+
+<p class='img'>
+  <img src='/images/blog/2016-07-reporting/libreoffice-import.png' />
+  Import of the CSV file 
+</p>
+
+After the import a graph can be created over the existing data.
+
+<p class='img'>
+  <img src='/images/blog/2016-07-reporting/libreoffice-graph.png' />
+  Graph in LibreOffice
+</p>
+
+For the generation of graphs without spreadsheet application is [matplotlib](http://matplotlib.org/) available. This is a powerful Python 2D plotting library. With the built-in support for SQLite in Python it will only take a couple lines of code to visualize your data. 
```

Comment:
You can also use matplotlib to generate graphs as an alternative to a spreadsheet application.